### PR TITLE
[Feature]: allow users to inject their own custom tracing layers

### LINKED ISSUE #1003: [Feature]: Let users inject their own tracing layers
### Describe the feature

As of right now, shuttle uses a [Registry](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/registry/struct.Registry.html) to pick up logs from [tracing](https://docs.rs/tracing/latest/tracing/). This allows for attaching multiple [Layer](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/layer/trait.Layer.html)s to the registry, which would then pick up the logs and can do whatever they want with them. Because shuttle sets up the registry, the user currently has no control over their own logs.

My idea was that we could add a parameter to the [#[shuttle_runtime::main]](https://docs.rs/shuttle-runtime/latest/shuttle_runtime/attr.main.html) macro that would allow the user to pass a function name that shuttle could then call to get one or more [Layer](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/layer/trait.Layer.html)s to attach to the registry.

The main problem with this approach is that [Layer](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/layer/trait.Layer.html) has a generic `S` which is the subscriber type it is attached to. I don't know if there is a way for the user to know that type ahead of time, so maybe we could expose a type alias from [shuttle_runtime](https://docs.rs/shuttle-runtime/latest/shuttle_runtime) that has the generic hard-coded according to whatever the default registry is?

### Suggestion or Example of how the feature would be used

The API for it could look something like this on the user side:

```rust
fn my_layer() -> Layer {
    let my_writer = todo!("...");

    tracing_subscriber::fmt::layer()
        .json()
        .with_writer(my_writer)
}

#[shuttle_runtime::main(tracing_layer = "my_layer")]
async fn init() -> shuttle_axum::ShuttleAxum {
    // ...
}
```

### Duplicate declaration

- [X] I have searched the issues and this feature has not been requested before.

### PR DESCRIPTION
**Pull request by AlphaKeks**
## Description of change
This adds the ability to pass a function into `#[shuttle_runtime::main]` that returns a tracing layer, which will then be attached to shuttle's tracing registry.

This aims to close #1003.

## How has this been tested? (if applicable)
- [x] Existing tests pass
- [x] Example ([currently here](https://github.com/AlphaKeks/shuttle-examples/tree/feat/custom-tracing-layer/tracing/custom-layer))
  - [ ] make it work without resorting to local paths?
- [x] Integration test to mock a custom logger and ensure sent logs being 
- [x] Make it easier for users to define their functions (maybe a custom type export?)

### CODE REVIEW

File: codegen/src/shuttle_main/mod.rs
Comment (jonaro00) [inline codegen/src/shuttle_main/mod.rs:320]: Errors should use `emit_error!` I assume?

**Review by jonaro00 — COMMENTED** [review]
Some comments. Would be good to rebase this to isolate changes.

File: codegen/src/shuttle_main/mod.rs
Comment (AlphaKeks) [reply codegen/src/shuttle_main/mod.rs:278]: No, it's not intended, I was under the impression each layer would have their own filtering in isolation. Should the addition of `filter_layer` just be delayed until after the injection?

Comment (jonaro00) [reply codegen/src/shuttle_main/mod.rs:278]: You might be on to something https://docs.rs/tracing-subscriber/0.3.17/tracing_subscriber/layer/index.html#per-layer-filtering

Comment (jonaro00) [reply codegen/src/shuttle_main/mod.rs:278]: Looks like you can apply with(filter) to each Layer individually instead of the the whole registry :)

**Comment by jonaro00** [comment]
To get rid of all the commits from main on this PR, you can `rebase -i` and drop all commits from main, that way only your changes should remain on this PR branch.

File: codegen/src/shuttle_main/mod.rs
Comment (AlphaKeks) [reply codegen/src/shuttle_main/mod.rs]: [Latest commit](https://github.com/shuttle-hq/shuttle/pull/1027/commits/d74b4b4f959106dfccce6a0ec06fe3a10d7102a5) should fix this.

**Comment by AlphaKeks** [comment]
> To get rid of all the commits from main on this PR, you can `rebase -i` and drop all commits from main, that way only your changes should remain on this PR branch.

Not sure how, would you mind explaining?

**Comment by jonaro00** [comment]
(fetch the main (upstream) branch), `git rebase -i upstream/main`, check the commit list here on github and let your own commits remain `pick`, while marking all others as `drop`. Then complete the rebase (hopefully no conflicts). Then force push.

**Comment by AlphaKeks** [comment]
Is there an easy way to write an integration test for this? Since this affects the `main` macro, the test would have to be its own binary that consumes the public API of the shuttle crates, but all the tests in `e2e/tests/integration` don't seem to do that. How would I go about writing a proper integration test for this?

**Comment by jonaro00** [comment]
We are thinking that you can keep building your example app in a way that makes it testable, such as a simple axum app that returns the logs that have been recorded so far. Then, an e2e test can be added and adapted to read that endpoint until a certain custom log line is found.
Does that sound good?

**Comment by AlphaKeks** [comment]
That's kind of a dependency cycle isn't it? Having an official example would require the functionality to already be merged, so the test could only really exist if the main repository already implements these changes.

I also don't think using an example for testing makes a lot of sense given that logging is something "internal" that no real API would ever expose? The purpose of the examples is to show users a practical approach of how to use shuttle, but I don't think anyone would make an API that returns some logs.

I'm happy to write tests for this, but I don't think this is the right approach.

**Comment by jonaro00** [comment]
> Having an official example would require the functionality to already be merged, so the test could only really exist if the main repository already implements these changes.

You don't need to worry much about the syncing of examples and main repo, the link is updated at every release. I think we could use a `next` branch on that repo for merging upcoming changes, but it has some side effects.

> I also don't think using an example for testing makes a lot of sense given that logging is something "internal" that no real API would ever expose? The purpose of the examples is to show users a practical approach of how to use shuttle, but I don't think anyone would make an API that returns some logs.

cargo-shuttle integration tests use the examples and run them, waiting for a correct response from their endpoints.
I agree that having an example that returns internal logs is not practical, but neither is one that prints the logs to stderr (no way of reading that when deployed). Ultimately, the only "practical" use is to send them to an external service like you showed on the demo (great demo btw), but this is too hard to incorporate in a test. So the suggestion above was my best compromise between testability and practicality. If you have a better suggestion, please let me know :)

*we literally have examples that return secrets haha. not very practical, but very testable ;)*

**Comment by AlphaKeks** [comment]
Should I just have the example in my own fork of the examples repo and depend on my `shuttle` fork via git / path?

**Comment by jonaro00** [comment]
I think there are multiple ways of getting it to work. If you find one of those, use it and we'll figure out something from there.

**Comment by AlphaKeks** [comment]
I wrote an integration test that hits the example API endpoint, collects some results and validates them. I couldn't figure out how to programmatically start the example from within the test, so if you could give me a hand there that would be great 😄 

The example is here: <https://github.com/AlphaKeks/shuttle-examples/commit/e9e23e1f1cda4d625f45786a2fb16ae482af3ee6>

It adds 2 endpoints; 1 to send messages and another one to read the last `n` send messages. The test simply sends some messages, then requests them and makes sure they are the right ones.

**Comment by jonaro00** [comment]
I'm thinking it can be done in a very similar fashion as `cargo-shuttle/tests/integration/run.rs`. Let me know if there are details that don't make sense.

**Comment by AlphaKeks** [comment]
I moved the `cargo_shuttle_run` function into a different crate so that it can be shared. The integration test now uses it to run the example, makes some requests, and validates them. The tests pass fine for me, although you need to checkout the correct branch inside the submodule for now. Shouldn't be an issue once the examples are merged into the main repository as well.

**Comment by jonaro00** [comment]
Great! Perhaps you can rebase away the conflicts and we can get to review and test this :D

**Comment by jonaro00** [comment]
You can open a PR to examples from your branch as well.

**Comment by AlphaKeks** [comment]
I'll add some documentation to it and open a PR!

**Review by jonaro00 — APPROVED** [review]
I moved the test to codegen/integration, as it technically isn't e2e, and now it can also run in CI. All looking good 🥳 
@oddgrd @chesedo Any final thoughts?

**Review by iulianbarbu — COMMENTED** [review]
LGTM so far, thanks! I'll just need to test this. I'll follow up soon.

File: codegen/tests/integration/custom_tracing_layer.rs
Comment (jonaro00) [reply codegen/tests/integration/custom_tracing_layer.rs:34]: Yeah, this test can take a more sequential/synchronous approach.

Comment (AlphaKeks) [reply codegen/tests/integration/custom_tracing_layer.rs:34]: The way the axum server works is that you make a GET request and it will wait for enough messages to come in before it responds, so you need to make the GET request first. The 500ms could probably be lower, I just chose a random value that worked consistently.

I couldn't think of a better way to make this work because it is, as already established, not very practical, and I just didn't come up with anything better than that 😅

Comment (jonaro00) [reply codegen/tests/integration/custom_tracing_layer.rs:34]: Alright, got a bit confused.

**Comment by jonaro00** [comment]
That is a great suggestion @robertohuertasm! It feels like a good idea to enable secret interpolation in that argument as well.

@AlphaKeks How did you implement this in your demo?

**Comment by AlphaKeks** [comment]
I didn't make the demo, but as far as I can tell @Kazy used environment variables for it.

I don't imagine it would be _too_ difficult to allow passing a [`SecretStore`](<https://docs.rs/shuttle-secrets/latest/shuttle_secrets/struct.SecretStore.html>) as well. We'd just need to check if the function provided by the user takes an argument or not, and make sure it has the correct type.

**Comment by Kazy** [comment]
Yes I used env variables, sorry ! On this topic, would you like me to push the demo to the `examples` repo ?

**Comment by robertohuertasm** [comment]
@Kazy but the env vars cannot be used in production, right?

**Comment by jonaro00** [comment]
I think that's the case. We probably want to allow secret interpolation in that part of the macro as well.

**Pull request merged**
